src/react/Reconciler.js

```diff
@@ -0,0 +1,332 @@
+            let diagnostic = new CompilerDiagnostic(
+              `__registerReactComponentRoot() failed due to - ${error.message}`,
+              this.realm.currentLocation,
+              "PP0019",
+              "FatalError"
+            );
+            this.realm.handleError(diagnostic);
+            throw new FatalError();
+          }
+          throw error;
+        }
+      })
+    );
+  }
+  _renderAsDeepAsPossible(
+    componentType: ECMAScriptSourceFunctionValue,
+    props: ObjectValue | AbstractValue,
+    context: ObjectValue | AbstractValue,
+    branchStatus: BranchStatusEnum
+  ) {
+    let { value, commitDidMountPhase, childContext } = this._renderOneLevel(
+      componentType,
+      props,
+      context,
+      branchStatus
+    );
+    let result = this._resolveDeeply(value, childContext, branchStatus);
+    return {
+      result,
+      childContext,
+      commitDidMountPhase,
+    };
+  }
+  _renderOneLevel(
+    componentType: ECMAScriptSourceFunctionValue,
+    props: ObjectValue | AbstractValue,
+    context: ObjectValue | AbstractValue,
+    branchStatus: BranchStatusEnum
+  ) {
+    if (valueIsClassComponent(this.realm, componentType)) {
+      // for now we don't support class components, so we throw a ExpectedBailOut
+      throw new ExpectedBailOut("class components not yet supported");
+    } else {
+      invariant(componentType.$Call, "Expected componentType to be a FunctionValue with $Call method");
+      let value = componentType.$Call(this.realm.intrinsics.undefined, [props, context]);
+      return { value, commitDidMountPhase: null, childContext: context };
+    }
+  }
+  _applyBranchedLogic(value: ObjectValue) {
+    // we need to apply a key when we're branched
+    let currentKeyValue = Get(this.realm, value, "key") || this.realm.intrinsics.null;
+    let uniqueKey = getUniqueReactElementKey("", this.reactSerializerState.usedReactElementKeys);
+    let newKeyValue = new StringValue(this.realm, uniqueKey);
+    if (currentKeyValue !== this.realm.intrinsics.null) {
+      newKeyValue = computeBinary(this.realm, "+", currentKeyValue, newKeyValue);
+    }
+    value.$Set("key", newKeyValue, value);
+    return value;
+  }
+  _updateBranchStatus(branchStatus: BranchStatusEnum): BranchStatusEnum {
+    if (branchStatus === BranchStatus.NEW_BRANCH) {
+      branchStatus = BranchStatus.BRANCH;
+    }
+    return branchStatus;
+  }
+  _resolveDeeply(value: Value, context: ObjectValue | AbstractValue, branchStatus: BranchStatusEnum) {
+    if (
+      value instanceof StringValue ||
+      value instanceof NumberValue ||
+      value instanceof BooleanValue ||
+      value instanceof NullValue ||
+      value instanceof UndefinedValue
+    ) {
+      // terminal values
+      return value;
+    } else if (value instanceof AbstractValue) {
+      for (let i = 0; i < value.args.length; i++) {
+        value.args[i] = this._resolveDeeply(value.args[i], context, BranchStatus.NEW_BRANCH);
```

Comment:
I think this is a bug. If both branches return the same component type, then we shouldn't append a unique key to it. It's only if the two branches have different component types that we should. Maybe add a test for this case:

```js
<div>{abstractVal ?<Foo arg={1} /> : <Foo arg={2} />}</div>
```

In this case state should be preserved.